org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/extension/SonarLintExtensionTracker.java

```diff
@@ -0,0 +1,143 @@
+import org.sonarlint.eclipse.core.SonarLintLogger;
+import org.sonarlint.eclipse.core.analysis.IAnalysisConfigurator;
+import org.sonarlint.eclipse.core.analysis.IFileLanguageProvider;
+import org.sonarlint.eclipse.core.configurator.ProjectConfigurator;
+import org.sonarlint.eclipse.core.resource.ISonarLintFileAdapterParticipant;
+import org.sonarlint.eclipse.core.resource.ISonarLintProjectFilter;
+import org.sonarlint.eclipse.core.resource.ISonarLintProjectsProvider;
+
+public class SonarLintExtensionTracker implements IExtensionChangeHandler {
+
+  private static final String ATTR_CLASS = "class"; //$NON-NLS-1$
+
+  private final SonarLintEP<ProjectConfigurator> configuratorEp = new SonarLintEP<>("org.sonarlint.eclipse.core.projectConfigurators"); //$NON-NLS-1$
+  private final SonarLintEP<IAnalysisConfigurator> analysisEp = new SonarLintEP<>("org.sonarlint.eclipse.core.analysisConfigurator"); //$NON-NLS-1$
+  private final SonarLintEP<ISonarLintProjectsProvider> projectsProviderEp = new SonarLintEP<>("org.sonarlint.eclipse.core.projectsProvider"); //$NON-NLS-1$
+  private final SonarLintEP<ISonarLintFileAdapterParticipant> fileAdapterParticipantEp = new SonarLintEP<>("org.sonarlint.eclipse.core.fileAdapterParticipant"); //$NON-NLS-1$
+  private final SonarLintEP<ISonarLintProjectFilter> projectFilterEp = new SonarLintEP<>("org.sonarlint.eclipse.core.projectFilter"); //$NON-NLS-1$
+  private final SonarLintEP<IFileLanguageProvider> languageEp = new SonarLintEP<>("org.sonarlint.eclipse.core.languageProvider"); //$NON-NLS-1$
+
+  private static class SonarLintEP<G> {
+
+    private final String id;
+    private final Collection<G> instances = new ArrayList<>();
+
+    public SonarLintEP(String id) {
+      this.id = id;
+    }
+  }
+
+  private final Collection<SonarLintEP<?>> allEps = Arrays.asList(configuratorEp, analysisEp, projectsProviderEp, fileAdapterParticipantEp, projectFilterEp, languageEp);
+
+  private ExtensionTracker tracker;
+
+  public void start() {
+    IExtensionRegistry reg = Platform.getExtensionRegistry();
+    tracker = new ExtensionTracker(reg);
+    IExtensionPoint[] epArray = allEps.stream().map(ep -> reg.getExtensionPoint(ep.id)).toArray(IExtensionPoint[]::new);
+    IFilter filter = ExtensionTracker.createExtensionPointFilter(epArray);
+    tracker.registerHandler(this, filter);
+    for (IExtensionPoint ep : epArray) {
+      for (IExtension ext : ep.getExtensions()) {
+        addExtension(tracker, ext);
+      }
+    }
+  }
+
+  public void close() {
+    if (tracker != null) {
+      tracker.close();
+      tracker = null;
+    }
+  }
+
+  @Override
+  public void addExtension(IExtensionTracker tracker, IExtension extension) {
+    IConfigurationElement[] configs = extension.getConfigurationElements();
+    for (final IConfigurationElement element : configs) {
+      try {
+        for (SonarLintEP ep : allEps) {
+          if (ep.id.equals(extension.getExtensionPointUniqueIdentifier())) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)